src/CommonLib/hostfxr_utility.cpp

```diff
@@ -429,105 +429,130 @@ HOSTFXR_UTILITY::FindDotnetExePath(
-        // however I'd rather error out here and report an issue.
-        hr = HRESULT_FROM_WIN32(ERROR_FILE_TOO_LARGE);
-        goto Finished;
-    }
+        // Check the bitness of the currently running process
+        // matches the dotnet.exe found. 
+        if (!IsWow64Process(GetCurrentProcess(), &fIsWow64Process))
+        {
+            // Calling IsWow64Process failed
+            hr = HRESULT_FROM_WIN32(GetLastError());
+            goto Finished;
+        }
+        if (fIsWow64Process)
+        {
+            // 32 bit mode
+            fIsCurrentProcess64Bit = FALSE;
+        }
+        else
+        {
+            SYSTEM_INFO systemInfo;
+            GetNativeSystemInfo(&systemInfo);
+            fIsCurrentProcess64Bit = systemInfo.wProcessorArchitecture == PROCESSOR_ARCHITECTURE_AMD64;
+        }
 
-    if (FAILED(hr = struDotnetLocationsString.CopyA(pzFileContents, dwNumBytesRead)))
-    {
-        goto Finished;
-    }
+        while (!fFound)
+        {
+            index = struDotnetLocationsString.IndexOf(L"\r\n", prevIndex);
+            if (index == -1)
+            {
+                break;
+            }
+            if (FAILED(hr = struDotnetSubstring.Copy(struDotnetLocationsString.QueryStr(), index - prevIndex)))
+            {
+                goto Finished;
+            }
+            prevIndex = index;
 
-    // Check the bitness of the currently running process
-    // matches the dotnet.exe found. 
-    if (!IsWow64Process(GetCurrentProcess(), &fIsWow64Process))
-    {
-        // Calling IsWow64Process failed
-        hr = HRESULT_FROM_WIN32(GetLastError());
-        goto Finished;
-    }
-    if (fIsWow64Process)
-    {
-        // 32 bit mode
-        fIsCurrentProcess64Bit = FALSE;
-    }
-    else
-    {
-        SYSTEM_INFO systemInfo;
-        GetNativeSystemInfo(&systemInfo);
-        fIsCurrentProcess64Bit = systemInfo.wProcessorArchitecture == PROCESSOR_ARCHITECTURE_AMD64;
+            if (GetBinaryTypeW(struDotnetSubstring.QueryStr(), &dwBinaryType) &&
+                fIsCurrentProcess64Bit == (dwBinaryType == SCS_64BIT_BINARY)) {
+                // Found a valid dotnet.
+                struDotnetPath->Copy(struDotnetSubstring);
+                fFound = TRUE;
+            }
+        }
     }
 
+Fallback:
+
+    // Look in ProgramFiles
     while (!fFound)
     {
-        index = struDotnetLocationsString.IndexOf(L"\r\n", prevIndex);
-        if (index == -1)
-        {
-            break;
-        }
-        if (FAILED(hr = struDotnetSubstring.Copy(struDotnetLocationsString.QueryStr(), index - prevIndex)))
+        if (FAILED(struDotnetSubstring.Resize(dwPathSize)))
         {
             goto Finished;
         }
-        prevIndex = index;
 
-        if (!GetBinaryTypeW(struDotnetSubstring.QueryStr(), &dwBinaryType))
+        // Program files will changes based on the 
+        dwNumBytesRead = GetEnvironmentVariable(L"ProgramFiles", struDotnetSubstring.QueryStr(), dwPathSize);
```

Comment:
You're only trying 64-bit fallback?

Reply:
Shouldn't it look in the appropriate place based on the bitness?

Reply:
%ProgramFiles% is bitness aware

Reply:
Ah I thought this was the path, I didn't see the GetEnvironmentVariable. Hopefully we have tests for this 😄

Reply:
Yeah as @Tratcher said, PF is bitness aware. 

@davidfowl native tests are coming soon ™️

Reply:
![image](https://user-images.githubusercontent.com/8302101/35581876-02e8c784-05a2-11e8-931a-a1f8a9926e63.png)
versus
![image](https://user-images.githubusercontent.com/8302101/35581905-12622912-05a2-11e8-9c26-998e40af2843.png)